proposals/p2875.md

```diff
@@ -0,0 +1,514 @@
+    more complex expression enclosed parentheses.
+-   `b`, `c`, `d` are any number of argument expressions. Arguments are
+    separated by commas, and if the argument list is not empty, an optional
+    trailing comma is permitted but not required after the final argument.
+
+Call syntax is syntactically equivalent to a primary expression followed by a
+tuple literal, except that a tuple literal requires a trailing comma to form a
+single-element tuple `(b,)`, whereas in call syntax both `a(b)` and `a(b,)` are
+permitted.
+
+### Direct calls
+
+A call expression is a _direct call_ when the callee:
+
+-   is the name of a parameterized entity, like a generic class or interface, or
+-   has a function type or bound method type.
+
+In a direct call, a call signature is available which is used to check the given
+arguments against the callee's declared implicit and explicit parameters. This
+checking proceeds as follows:
+
+-   Argument deduction is performed by comparing the declared parameter types
+    against the actual parameter types and deducing values for implicit
+    arguments that make the types equal.
+-   Then, for each binding in the explicit parameter list in turn, all argument
+    values that have been deduced are substituted into the parameter.
+
+    -   If the parameter is a `template :!` binding, the argument expression is
+        converted to a constant value phase expression of the type of the
+        binding.
+    -   If the parameter is a checked `:!` binding, the argument expression is
+        converted to a symbolic value phase expression of the type of the
+        binding.
+    -   Otherwise, the parameter is pattern-matched against the argument.
+
+    If a parameter is a `:!` binding, its corresponding converted argument
+    expression is evaluated, and its value is added to the list of deduced
+    argument values before any later parameters are processed.
+
+The result of the call expression depends on the callee:
+
+-   If the callee is a parameterized entity such as a generic class or a generic
+    interface, the result is the specific instance of that generic, such as a
+    class or interface, and the call is a value expression of type `type`.
+-   If the callee is a function value, the call is an initializing expression
+    whose type is the substituted return type of the function. When evaluated,
+    the call expression will invoke the function and produce whatever value it
+    returns.
+-   If the callee is a bound method value, it behaves the same as a function
+    value, except that the `self` parameter of the called function is bound to
+    the `self` value in the bound method value.
+
+### Generic callable parameters
+
+A generic parameter can be constrained to be a callable type using the `Call`
+interface:
+
+```
+interface Call(Args: type) {
+  let Result:! type;
+  fn Op[self: Self](args: Args) -> Result;
+}
+```
```

Comment:
I feel like we need 2-3 variations of this interface with different `self` parameters, with an `extend` relationship between them.

Reply:
Yeah, I think that's a good point, that we shouldn't forget about. I've added a section to "Future work" on this; I think we can decouple that from this proposal and add the extra interfaces separately.